draft-ietf-httpbis-message-signatures.md

```diff
@@ -237,22 +237,26 @@ The following sections define component identifier names, their parameters, thei
 
 ## HTTP Fields {#http-fields}
 
-The component name for an HTTP field is the lowercased form of its field name. While HTTP field names are case-insensitive, implementations MUST use lowercased field names (e.g., `content-type`, `date`, `etag`) when using them as component names.
+The component name for an HTTP field is the lowercased form of its field name as defined in {{Section 5.1 of HTTP}}. While HTTP field names are case-insensitive, implementations MUST use lowercased field names (e.g., `content-type`, `date`, `etag`) when using them as component names.
 
-Unless overridden by additional parameters and rules, the HTTP field value MUST be canonicalized as a single combined value as defined in {{Section 5.2 of HTTP}}. Note that some HTTP fields, such as Set-Cookie {{COOKIE}}, do not follow a syntax that allows for combination of field values in this manner such that the combined output is unambiguous from multiple inputs. However, the canonicalized component value is never parsed by the message signature process; instead, it is merely used as part of the signature base in {{create-sig-input}}. Even so, caution needs to be taken when including such fields in signatures, and the `bs` parameter defined in {{http-header-byte-sequence}} provides a method for wrapping such problematic fields. See {{security-non-list}} for more discussion of this issue.
+The component value for an HTTP field is the field value for the named field as defined in {{Section 5.5 of HTTP}}. The field value MUST be taken from the named header field of the target message unless this behavior is overridden by additional parameters and rules, such as the `req` and `tr` flags, below.
 
-If the combined value is not available for a given field by an implementation, the following algorithm will produce canonicalized results for an implementation:
+Unless overridden by additional parameters and rules, list-based HTTP field values MUST be combined into a single value as defined in {{Section 5.2 of HTTP}}. Specifically, list-based HTTP fields sent as multiple fields MUST be combined using a single comma (",") and a single space (" ") between each item. Note that intermediaries are allowed to separate values of list-based HTTP fields with any amount of whitespace between commas. If this behavior is not accounted for by the verifier, the signature can fail since the addition or removal of spaces between list items will change the signature base. It is RECOMMENDED that signers and verifiers process list-based fields starting with the individual field values based on the strict algorithm below, where possible, to account for possible intermediary behavior.
 
```

Comment:
I believe the only robust way to sign a field is to actually make sure that only one instance of the field is present in the signed message (thus canonicalize not only the signed value, but the actual message).

Reply:
@reschke done, please check the new text to see if it aligns with your point

Reply:
There are stimm multiple issues here:

- this not just about "list based" fields, but any case where there are multiple instances of the same field name - that may soundlike nitpicking, but the combination rule applies to *any* field
- "Note that intermediaries are allowed to separate values of list-based HTTP fields with any amount of whitespace between commas." - that sounds a bit like if this was about uncombining lists, but that's no whay you mean here , right?
- "For robustness, it is RECOMMENDED that messages include only a single value for any field covered under the signature, with the value for list-based fields serialized using the algorithm below." - once it's a single value, the actua algorithm doesn't matter at all.
- "This approach increases the chances of the field value remaining untouched through intermediaries." - it sort of guarantees it, unless the intermediary rewrites values it shouldn't touch (in which case we want the sig to fail)
- "Where that approach is not possible and multiple field lines need to be sent separately, it is RECOMMENDED that signers and verifiers process list-based fields starting with the individual field values based on the strict algorithm below, where possible, to account for possible intermediary behavior." - that makes it sound as if that would help - but if the intermediary combines in a different way (such as no SP), the sig will definitively fail.

Reply:
No, intermediaries are not supposed to rewrite field values. In particular, an intermediary cannot assume that a comma is indeed a list delimiter in general -- consider quoted strings. Or ETags.

Reply:
I've taken another shot at this -- please take a look at the new text.